File: vm/atom.rb
class VM
  class Atom < String
    attr_reader :filename, :offset, :line, :column

    def initialize(name, filename = nil, line = nil, column = nil)
      super(name.to_s)
      @filename = filename
      @line = line
      @column = column
    end

    def raw
      self
    end

    alias eq? ==

    def to_ruby
      to_s
    end

    def mangle(version)
      name = "##{self}.v#{version}"
      self.class.new(name, filename, line, column)
    end
  end
end

File: src/atom.rs
use rb;
use rb::{Value, RB_NIL};

pub fn atom(name: &str, filename: &str, offset: usize, newlines: &Vec<usize>) -> Value {
    let lines_before: Vec<usize> = newlines.iter().take_while(|i| *i < &offset).map(|i| *i).collect();
    let line = int2rbnum!(lines_before.len() + 1);
    let column = match lines_before.last() {
        Some(i) => int2rbnum!(offset - i),
        None => int2rbnum!(offset + 1)
    };
    let name = rb::str_new(&name.to_string());
    let filename = rb::str_new(&filename.to_string());
    let vm = rb::const_get("VM", &RB_NIL);
    let atom = rb::const_get("Atom", &vm);
    rb::class_new_instance(&atom, vec![name, filename, line, column])
}
